modAionImpl/src/org/aion/zero/impl/sync/msg/ResponseTrieData.java

```diff
@@ -0,0 +1,259 @@
+     * @param dbType the blockchain database in which the key should be found
+     * @throws NullPointerException if any of the given parameters are {@code null}
+     */
+    public ResponseTrieData(
+            final ByteArrayWrapper nodeKey, final byte[] nodeValue, final DatabaseType dbType) {
+        super(Ver.V1, Ctrl.SYNC, Act.RESPONSE_TRIE_DATA);
+
+        // ensure inputs are not null
+        Objects.requireNonNull(nodeKey);
+        Objects.requireNonNull(nodeValue);
+        Objects.requireNonNull(dbType);
+
+        this.nodeKey = nodeKey;
+        this.nodeValue = nodeValue;
+        this.referencedNodes = new HashMap<>();
+        this.dbType = dbType;
+    }
+
+    /**
+     * Constructor for trie node responses that contain multiple nodes (key-value pairs).
+     *
+     * @param nodeKey the key of the requested trie node
+     * @param nodeValue the value stored for the requested trie node
+     * @param referencedNodes a map of key-value pairs referenced by the value of the requested key
+     * @param dbType the blockchain database in which the key should be found
+     * @throws NullPointerException if any of the given parameters are {@code null}
+     * @implNote The referenced nodes are purposefully not deep copied to minimize resource
+     *     instantiation. This is a reasonable choice given that these objects are created to be
+     *     encoded and transmitted over the network and therefore there is the expectation that they
+     *     will not be utilized further.
+     */
+    public ResponseTrieData(
+            final ByteArrayWrapper nodeKey,
+            final byte[] nodeValue,
+            final Map<ByteArrayWrapper, byte[]> referencedNodes,
+            final DatabaseType dbType) {
+        super(Ver.V1, Ctrl.SYNC, Act.RESPONSE_TRIE_DATA);
+
+        // ensure inputs are not null
+        Objects.requireNonNull(nodeKey);
+        Objects.requireNonNull(nodeValue);
+        Objects.requireNonNull(dbType);
+        Objects.requireNonNull(referencedNodes);
+
+        this.nodeKey = nodeKey;
+        this.nodeValue = nodeValue;
+        this.referencedNodes = referencedNodes;
+        this.dbType = dbType;
+    }
+
+    /**
+     * Decodes a message into a trie node response.
+     *
+     * @param message a {@code byte} array representing a response to a trie node request.
+     * @return the decoded trie node response
+     * @implNote The decoder must return {@code null} if any of the component values are {@code
+     *     null} or invalid.
+     */
+    public static ResponseTrieData decode(final byte[] message) {
+        if (message == null || message.length == 0) {
+            return null;
+        } else {
+            RLPList list = (RLPList) RLP.decode2(message).get(0);
+            if (list.size() != 4) {
+                return null;
+            } else {
+                // decode the key
+                byte[] hash = list.get(0).getRLPData();
+                if (hash.length != 32) {
```

Comment:
might be worth putting 4, 32 in consts since they seem to show up all over the place

Reply:
I added `org.aion.p2p.V1Constants` for this purpose.